tests/models/test_restore.py

```diff
@@ -10,11 +10,52 @@
 
 import tests.base.develop_pipelines as tpipes
 import tests.base.develop_utils as tutils
-from pytorch_lightning import Trainer
+from pytorch_lightning import Trainer, Callback
 from pytorch_lightning.callbacks import ModelCheckpoint
 from tests.base import EvalModelTemplate, GenericEvalModelTemplate
 
 
+class ModelTrainerPropertyParity(Callback):
+
+    def check_properties(self, trainer, pl_module):
```

Comment:
```suggestion
    def _check_properties(self, trainer, pl_module):
```

Reply:
Basically, before the fix, if we resume the training with a particular checkpoint (epoch>1), the following assertions fail. 
`assert trainer.global_step == pl_module.global_step`
`assert trainer.current_epoch == pl_module.current_epoch`

This is because `pl_module.global_step` or `pl_module.current_epoch` properties are not the same as that of `trainer` at different stages. To make sure the same, the test function was added by Adrian.

Reply:
This probably comes from `self.increment_accumulated_grad_global_step()` of training_loop.py.

Reply:
@ydcjeff are you saying there is another bug?
the bug here is simply that when resuming from checkpoint, the model.global_step get's set only in the training loop. This is too late, since you may need it in configure optimizer or other hook that runs right after model is init. I think it make sense to have model grab the global step directly from trainer instead of keeping a separate count. 
Also, by introducing the property here, we make sure the user does not accidentally overrwrite the global step or current epoch in their LightningModule, because if they don't know about it they might try to track their own variables and this would overwrite ours. This is my main concern actually also with the other properties, they are trainer properties duplicated into the LightningModule but there is not guarantee for parity throughout the training loop.

Reply:
ahhh no no I don't mean like that.
sorry @awaelchli if i confuse you 😓, i just point out where trainer.global_step get incremented.
I will be careful with my comment.

Reply:
@ydcjeff everything is good! don't worry. You don't need to be careful about anything. Maybe my text above was a bit long and intimidating. But the intention was the opposite. I like to explain things in order to understand and give others the chance to disagree with me so I can figure it out. Basically my question was: 
Does the above logic make sense or do we have another bug connected to accumulate gradients?

Reply:
Yes, it makes sense.